src/librustdoc/html/render/search_index.rs

```diff
@@ -28,6 +31,52 @@ use crate::formats::item_type::ItemType;
 use crate::html::markdown::short_markdown_summary;
 use crate::html::render::{self, IndexItem, IndexItemFunctionType, RenderType, RenderTypeId};
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct CratePath(ItemType, Vec<Symbol>);
+
+impl Hash for CratePath {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // We have to manually implement this because we're not allowed to rely on the (current)
+        // implementation details of how collections are hashed
+        self.0.hash(state);
+        state.write_length_prefix(self.1.len());
+        for sym in &self.1 {
+            sym.hash(state);
+        }
+    }
+}
+
+/// This struct allows doing lookups on a map where the keys are [`CratePath`]s, without having to allocate a vector.
+struct CratePathRef<'sym>(ItemType, &'sym [&'sym [Symbol]]);
+
+impl<'sym> CratePathRef<'sym> {
+    fn symbols(&self) -> impl Iterator<Item = &Symbol> {
+        self.1.iter().copied().flatten()
+    }
+}
+
+impl<'a, 'sym> From<&'a CratePathRef<'sym>> for CratePath {
+    fn from(value: &'a CratePathRef<'sym>) -> Self {
+        Self(value.0, value.symbols().copied().collect())
+    }
+}
+
+impl<'sym> Equivalent<CratePath> for CratePathRef<'sym> {
+    fn equivalent(&self, key: &CratePath) -> bool {
+        self.0 == key.0 && self.symbols().eq(&key.1)
+    }
+}
+
+impl<'sym> Hash for CratePathRef<'sym> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
+        state.write_length_prefix(self.symbols().count());
+        for sym in self.symbols() {
+            sym.hash(state);
+        }
```

Comment:
If CratePath::hash just calls CratePathRef::hash instead of reimplementing it, then you don't need to do your own Hash impl, and can do this (or even derive(Hash), I think).

```suggestion
        self.symbols()[..].hash(state);
```